src/main/scala/lms/core/codegen/cpp_codegen.scala

```diff
@@ -0,0 +1,102 @@
+package lms.core
+
+import lms.macros.RefinedManifest
+
+import scala.collection.mutable
+
+import java.io.{ByteArrayOutputStream, PrintStream}
+
+import Backend._
+import lms.core.stub._
+
+trait CPPOps { b: Base =>
+
+  abstract class Capture
+  object Capture {
+    def apply(s: String) = Seq(Share(s))
+    def CopyExcept[T:Manifest](s: Rep[T]*) = Seq(Share("=")) ++ s.map(UniqueRef(_))
+    def RefExcept[T:Manifest](s: Rep[T]*) = Seq(Share("&")) ++ s.map(UniqueCopy(_))
+  }
+  case class Share(s: String) extends Capture
+  case class UniqueCopy[T:Manifest](s: Rep[T]) extends Capture
+  case class UniqueRef[T:Manifest](s: Rep[T]) extends Capture
```

Comment:
I think I can guess the meaning of `Share`, `UniqueCopy` and `UniqueRef`, but It seems not an intuitive mapping to conventions (by-copy and by-ref) unless they have other meanings. Especially, what does "unique" mean here? they remind me shared_ptr and unique_ptr which is confusing.

Reply:
the user is not supposed to use them, but rather use the functions in `object Capture`.

I tried to make the case classes `private` but it doesn't compile :)

Reply:
I think he meant for us the programmer. The fact that users don't have access doesn't mean we shouldn't give them name that makes sense.

Reply:
fixed :)